lib/SyTest/Federation/Server.pm

```diff
@@ -96,6 +103,34 @@ sub get_event
    return $event;
 }
 
+sub get_auth_chain
+{
+   my $self = shift;
+   my @event_ids = @_;
+
+   my %events_by_id = map { $_ => $self->get_event( $_ ) } @event_ids;
+
+   my @auth_chain = @event_ids;
+
+   while( @event_ids ) {
+      my $event = $events_by_id{shift @event_ids};
+
+      my @auth_ids = map { $_->[0] } @{ $event->{auth_events} };
+
+      foreach my $id ( @auth_ids ) {
+         next if $events_by_id{$id};
+
+         $events_by_id{$id} = $self->get_event( $id );
+         push @event_ids, $id;
+      }
+
+      unshift @auth_chain, @auth_ids;
+   }
+
+   # uniqify by first occurence
```

Comment:
It wouldn't quite be, because here I'm starting at the "most recent" events and walking backwards chronologically via each implied earlier event, building the full auth_chain list backwards. I then uniq it by first occurrence, ensuring that no event appears before the things that would authenticate it; as would happen were it uniq'ed as it went.

Though I suppose it might be nicer to do

    @auth_chain = uniq @auth_ids, @auth_chain;

incrementally here at least.